src/lib/zkapp.ts

```diff
@@ -489,70 +488,61 @@ class Callback<Result> extends GenericArgument {
   methodIntf: MethodInterface & { returnType: Provable<Result> };
   args: any[];
 
+  result?: Result;
+  accountUpdate: AccountUpdate;
+
   static create<T extends SmartContract, K extends keyof T>(
     instance: T,
     methodName: K,
     args: T[K] extends (...args: infer A) => any ? A : never
   ) {
-    let callback: Callback<T[K] extends (...args: any) => infer R ? R : never> =
-      new this(instance, methodName, args);
-    return callback;
-  }
-
-  private constructor(instance: any, methodName: any, args: any[]) {
-    super();
-    this.instance = instance;
     let ZkappClass = instance.constructor as typeof SmartContract;
-    let methodIntf = (ZkappClass._methods ?? []).find(
+    let methodIntf_ = (ZkappClass._methods ?? []).find(
       (i) => i.methodName === methodName
     );
-    if (methodIntf === undefined)
+    if (methodIntf_ === undefined)
       throw Error(
-        `Callback: could not find method ${ZkappClass.name}.${methodName}`
+        `Callback: could not find method ${ZkappClass.name}.${String(
+          methodName
+        )}`
       );
-    methodIntf = {
-      ...methodIntf,
-      returnType: methodIntf.returnType ?? provable(null),
+    let methodIntf = {
+      ...methodIntf_,
+      returnType: methodIntf_.returnType ?? provable(null),
     };
-    this.methodIntf = methodIntf as any;
-    this.args = args;
+
+    // call the callback, leveraging composability (if this is inside a smart contract method)
+    // to prove to the outer circuit that we called it
+    let result = (instance[methodName] as Function)();
+    let accountUpdate = instance.self;
+
+    let callback = new Callback<any>({
+      instance,
+      methodIntf,
+      args,
+      result,
+      accountUpdate,
+      isEmpty: false,
+    });
+
+    return callback;
+  }
+
+  private constructor(self: Callback<any>) {
+    super();
+    Object.assign(this, self);
   }
 }
 
-// TODO: prove call signature in the outer circuit, just like for composability!
 function accountUpdateFromCallback(
```

Comment:
Sorry to be that guy in the PR but could we add some doc comments here since we are modifying the function? :D

Reply:
Thanks for being that guy :D Since `accountUpdateFromCallback` was the almost the same as `authorize` anyway, instead of doc-commenting it I ended up removing it and making `authorize` also take a callback. API feels cleaner now. I also added doc-comments to `authorize`!